napari/utils/migrations.py

```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
     _OldClass.__new__.__signature__ = prealloc_signature  # type: ignore [attr-defined]
 
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
+    """A dictionary that issues warning messages when deprecated keys are accessed.
+
+    Deprecated keys and values are not stored as part of the dictionary, so will not
+    appear when iterating over this or its items.
+
+    Instead deprecated items can only be accessed using `__getitem__`, `__setitem__`,
+    and `__delitem__`, or using `self.deprecations` directly.
+    """
+
+    # Maps from a deprecated key to its value and deprecation message.
+    deprecations: dict[str, tuple[Any, str]]
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.deprecations = {}
+
+    def __getitem__(self, key: str) -> Any:
+        if key in self.deprecations:
+            value, message = self.deprecations[key]
+            warnings.warn(message, FutureWarning)
+            return value
```

Comment:
```suggestion
            key, message = self._deprecations[key]
            warnings.warn(message, FutureWarning)
```
To enforce the same value for base and deprecated argument

Reply:
In https://github.com/napari/napari/pull/5477 I'll need a mechanism to deprecate things that are not just renames, so this is not sufficient.

We could add a function to the deprecated value to transform the base/non-deprecated value, but then we'd also need a setter. In the future, even this behavior might be limited, so I'd want a more general way to store deprecated values that are not easily derivable from the base/non-deprecated values.